### PR TITLE
build: migrate clang-tidy to LLVM 16

### PR DESCRIPTION
**Pull request by scarf005**
#### Summary

SUMMARY: Infrastructure "DDA Port: migrate clang-tidy plugin to LLVM 16"

#### Purpose of change

Make it easier to get custom clang-tidy plugin easier on local environment.

#### Describe the solution

ported following PRs: 

- https://github.com/CleverRaven/Cataclysm-DDA/pull/65381
- https://github.com/CleverRaven/Cataclysm-DDA/pull/56506
- https://github.com/CleverRaven/Cataclysm-DDA/pull/65806

all credits to: @BrettDong @jbytheway @Qrox 

#### Describe alternatives you've considered

RIIR

#### Testing

image

confimed it was possible to use plugin built with local LLVM 16. CI may need to be adjusted a lot since i've patchworked a lot of files.

also lots of new checks are disabled as they were too time consuming and killed runners due to 6h limit. will be handled in separate PRs.

### CODE REVIEW

**Comment by Qrox** [comment]
You might want to also apply CleverRaven/Cataclysm-DDA#66315 and CleverRaven/Cataclysm-DDA#66538 which are follow-ups to CleverRaven/Cataclysm-DDA#65806.

**Comment by scarf005** [comment]
thank you! will try tomorrow

File: build-scripts/build.sh [hunk 1]
```diff
@@ -95,15 +96,52 @@ then
         # And the same for clang-tidy
         "$CATA_CLANG_TIDY" ../src/version.cpp -- -v
 
+        # Run clang-tidy analysis instead of regular build & test
+        # We could use CMake to create compile_commands.json, but that's super
+        # slow, so use compiledb <https://github.com/nickdiego/compiledb>
+        # instead.
+        compiledb -n make
```
Comment (olanti-p) [inline build-scripts/build.sh:103]: `compiledb` shouldn't be needed anymore, see #2474

File: .github/workflows/clang-tidy.yml [hunk 2]
```diff
@@ -1,12 +1,21 @@
...
+    - '**.c'
+    - '**.cpp'
+    - '**.h'
+    - '**.hpp'
+    - '**/CMakeLists.txt'
+    - '**/Makefile'
+    - 'build-scripts/clang-tidy.sh'
+    - 'build-scripts/clang-tidy-wrapper.sh'
+    - 'build-scripts/get_affected_files.py'
+    - '.github/workflows/clang-tidy.yml'
```
Comment (olanti-p) [inline .github/workflows/clang-tidy.yml:18]: Why remove `tools/clang-tidy-plugin/` and others?

File: .github/workflows/clang-tidy.yml [hunk 3]
```diff
@@ -25,36 +34,78 @@ jobs:
         uses: fkirc/skip-duplicate-actions@master
         with:
           cancel_others: "true"
-          paths: '[ "**.cpp", "**.h", "**.c", "**/CMakeLists.txt", "**/Makefile", "**.hpp", "**.cmake", "build-scripts/**", "tools/clang-tidy-plugin/**", ".github/workflows/clang-tidy.yml", "**/.clang-tidy" ]'
+          paths: '[ "**.cpp", "**.h", "**.c", "**/CMakeLists.txt", "**/Makefile", "**.hpp", "**.cmake", "build-scripts/clang-tidy.sh", "build-scripts/clang-tidy-wrapper.sh", "build-scripts/get_affected_files.py", ".github/workflows/clang-tidy.yml" ]'
```
Comment (olanti-p) [inline .github/workflows/clang-tidy.yml:37]: Same here

File: build-scripts/clang-tidy.sh [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+# Show compiler C++ header search path
+${COMPILER:-clang++} -v -x c++ /dev/null -c
+# And the same for clang-tidy
+"$CATA_CLANG_TIDY" ../src/version.cpp -- -v
+
+# Run clang-tidy analysis instead of regular build & test
+# We could use CMake to create compile_commands.json, but that's super
+# slow, so use compiledb <https://github.com/nickdiego/compiledb>
+# instead.
+compiledb -n make
```
Comment (olanti-p) [inline build-scripts/clang-tidy.sh:73]: Same here, at line 33 cmake is invoked with `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON`

File: tools/clang-tidy-plugin/HeaderGuardCheck.cpp [hunk 5]
```diff
@@ -50,7 +49,9 @@ static std::string getHeaderGuard( StringRef Filename )
     bool Found = false;
     while( std::string::npos != ( LastSlash = TopDir.find_last_of( "/\\" ) ) ) {
         TopDir = TopDir.substr( 0, LastSlash );
-        if( pathExists( TopDir + "/.astylerc" ) ) {
+        // Either the root source dir (containing .travis.yml) or the root build
+        // dir (containing CMakeCache.txt)
+        if( pathExists( TopDir + "/.travis.yml" ) || pathExists( TopDir + "/CMakeCache.txt" ) ) {
```
Comment (olanti-p) [inline tools/clang-tidy-plugin/HeaderGuardCheck.cpp:54]: We don't have travis anymore, this part should be reverted

File: tools/clang-tidy-plugin/JsonTranslationInputCheck.cpp [hunk 6]
```diff
@@ -30,9 +30,7 @@ void JsonTranslationInputCheck::registerMatchers( MatchFinder *Finder )
                             anyOf(
                                 functionDecl(
                                     hasAnyName(
-                                        "_", "translation_argument_identity", "gettext", "pgettext",
-                                        "n_gettext", "npgettext"
-                                    )
+                                        "_", "translation_argument_identity", "pgettext" )
```
Comment (olanti-p) [inline tools/clang-tidy-plugin/JsonTranslationInputCheck.cpp]: Why remove? We have `vgettext` and `vpgettext` as alternatives to `n_gettext` and `npgettext`.

File: tools/clang-tidy-plugin/NoStaticTranslationCheck.cpp [hunk 7]
```diff
@@ -19,19 +19,12 @@ void NoStaticTranslationCheck::registerMatchers( MatchFinder *Finder )
                 decl(
                     anyOf(
                         functionDecl(
-                            hasAnyName( "_", "translation_argument_identity", "gettext", "pgettext",
-                                        "n_gettext", "npgettext" )
+                            hasAnyName( "_", "translation_argument_identity", "gettext", "pgettext" )
```
Comment (olanti-p) [inline tools/clang-tidy-plugin/NoStaticTranslationCheck.cpp]: Same here.

File: tools/clang-tidy-plugin/TranslateStringLiteralCheck.cpp [hunk 8]
```diff
@@ -31,9 +31,7 @@ void TranslateStringLiteralCheck::registerMatchers( MatchFinder *Finder )
                                             "_",
                                             "translation_argument_identity",
                                             "gettext",
-                                            "pgettext",
-                                            "n_gettext",
-                                            "npgettext"
+                                            "pgettext"
```
Comment (olanti-p) [inline tools/clang-tidy-plugin/TranslateStringLiteralCheck.cpp]: Same here.

File: tools/clang-tidy-plugin/TranslatorCommentsCheck.cpp [hunk 9]
```diff
@@ -236,14 +236,6 @@ void TranslatorCommentsCheck::registerMatchers( MatchFinder *Finder )
...
         this
     );
-    Finder->addMatcher(
-        callExpr(
-            callee( functionDecl( hasName( "n_gettext" ) ) ),
-            hasImmediateArgument( 0, stringLiteralArgumentBound ),
-            hasImmediateArgument( 1, stringLiteralArgumentUnbound )
-        ),
-        this
-    );
```
Comment (olanti-p) [inline tools/clang-tidy-plugin/TranslatorCommentsCheck.cpp]: And here.

File: tools/clang-tidy-plugin/TranslatorCommentsCheck.cpp [hunk 10]
```diff
@@ -269,15 +261,6 @@ void TranslatorCommentsCheck::registerMatchers( MatchFinder *Finder )
...
     );
-    Finder->addMatcher(
-        callExpr(
-            callee( functionDecl( hasAnyName( "npgettext" ) ) ),
-            hasImmediateArgument( 0, stringLiteralArgumentUnbound ),
-            hasImmediateArgument( 1, stringLiteralArgumentBound ),
-            hasImmediateArgument( 2, stringLiteralArgumentUnbound )
-        ),
-        this
-    );
```
Comment (olanti-p) [inline tools/clang-tidy-plugin/TranslatorCommentsCheck.cpp]: And here.

File: tools/clang-tidy-plugin/test/translator-comments.cpp [hunk 11]
```diff
@@ -27,13 +27,6 @@ void foo()
     //~ bar
     _( "bar" );
 
-    vgettext( /*~ foo */ ( "bar" ), _( "baz" ), 0 );
-
-    //~ bar
-    gettext( "bar" );
-
-    //~bar
-    vgettext( "bar", "baz", 1 );
```
Comment (olanti-p) [inline tools/clang-tidy-plugin/test/translator-comments.cpp:79]: And here.

File: tools/clang-tidy-plugin/test/translator-comments.cpp [hunk 12]
```diff
@@ -77,16 +70,6 @@ void foo()
...
-    /*~
-    barnacle */
-    vpgettext( "foo",
-               "bar", "baz", 2 );
-    // CHECK-MESSAGES: [[@LINE-4]]:5: warning: Translator comment without a matching raw string
-
-    //~ barn
-    vpgettext( "foo", "bar", ( "baz" ), 3 );
-    // CHECK-MESSAGES: [[@LINE-2]]:5: warning: Translator comment without a matching raw string
-
```
Comment (olanti-p) [inline tools/clang-tidy-plugin/test/translator-comments.cpp:132]: And here.

**Review by olanti-p — CHANGES_REQUESTED** [review]
Some checks definitely need to be disabled. The tidy runner seems to be timing out because of the sheer amount of new warnings.

File: .github/workflows/clang-tidy.yml (hunk 2 above)
Comment (scarf005) [reply .github/workflows/clang-tidy.yml:18]: must've overwritten accidentally.

**Comment by scarf005** [comment]
@olanti-p maybe we could merge the PR, turn off clang-tidy lint action for a week while opening PRs with `clang-tidy --fix` applied? maybe after merging #2216 and #2250?

**Comment by olanti-p** [comment]
Why disable entire clang-tidy action, if you can just disable the offending lints?

**Comment by scarf005** [comment]
tidy was running on every source file, causing timeout.

**Comment by scarf005** [comment]
> tidy was running on every source file, causing timeout. that said, i should fix that.

and i can't still figure out how.

**Comment by scarf005** [comment]
> and i can't still figure out how.

found out the curprit, fixed in [`516071e` (#3028)](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/3028/commits/516071e4931c3e7ce0fa1aa91aceadd529874320)

**Comment by scarf005** [comment]
temporarily disabled 'check all affected files' for now (three header changed in this PR affects entire .cpp)

**Review by olanti-p — CHANGES_REQUESTED** [review]
It doesn't seem to work at all now.

**Comment by scarf005** [comment]
aww, fixed. do you think it'd be ok to temporarily disable checking all affected source files and only run clang-tidy on changed files?

**Comment by olanti-p** [comment]
If you mean, disable checking `.cpp`s if they were not changed but include a header that was changed, then no, it won't work this way.

C++ compilers (and clang-tidy checks by extension) need the entire preprocessed file to work with, and preprocessed file is created from `.cpp` and all its `#include`d headers + defines passed to the compiler.

If that "affected files" check was disabled, then one PR could introduce error in header file that's only apparent when analyzing together with relevant .cpp, but CI would accept that, and then some other PR that happens to (correctly) modify that .cpp in a different place would trigger the check and get yelled at for no reason.

**Comment by olanti-p** [comment]
As I've mentioned before, you should start with fixing and/or disabling the checks added in this PR. If you check the build logs, the new checks spew billions upon billions of repeating errors, so it's quite possible the build times out simply because clang spends most of the time formatting error output.

BN builds 2-3 times faster than DDA, I say it's safe to expect our tidy could also run 2-3 times faster.

**Comment by scarf005** [comment]
image

a few days ago i tried disabling all new checks and still stumbled across 360m limit, as `point.h`, `make_static.h`, `translations.h` affected entire codebase. let's try again...

**Comment by scarf005** [comment]
every new checks are disabled but it still hits time limit. this time limit hit is specifically caused due to [changes in aforementioned 3 headers](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/3028#issuecomment-1747992630). which caused clang-tidy to run checks for every source files.

i think this PR should be merged despite that, as 
- [affected file tracing behavior is kept back](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/3028/commits/c12cf758e15191dccb2de4904934f74fdcd79b5e)
- it's planned to open followup PRs with running clang-tidy.

**Comment by olanti-p** [comment]
I strongly disagree.

Tidy job works fine, while this PR in its current state breaks it completely.

Running tidy locally should not be a priority over running it in CI, because nobody runs it locally except when they're working on some tidy issue, but CI runs it for all C++ changes.

There's no immediate need to upgrade to tidy + LLVM 16, this PR is not a blocker for anything, it should be a self-contained upgrade, not downgrade.

If you claim you've disabled all the new checks and still it times out, takes more than 6 hours to complete, but at the current state of things tidy works and takes less than 2 hours _for the entire run on all files_, where does the >3 times increase come from?

**Comment by scarf005** [comment]
keeping it draft until #2250 to reduce cognitive load

**Pull request closed without merging**